flang/lib/Lower/Bridge.cpp

```diff
@@ -801,161 +798,255 @@ class FirConverter : public Fortran::lower::AbstractConverter {
   ///  - structured and unstructured increment loops
   ///  - structured and unstructured concurrent loops
   void genFIR(const Fortran::parser::DoConstruct &) {
+    // Collect loop information.
+    // Generate begin loop code directly for infinite and while loops.
     auto &eval = getEval();
     bool unstructuredContext = eval.lowerAsUnstructured();
-    Fortran::lower::pft::Evaluation &doStmtEval =
-        eval.getFirstNestedEvaluation();
+    auto &doStmtEval = eval.getFirstNestedEvaluation();
     auto *doStmt = doStmtEval.getIf<Fortran::parser::NonLabelDoStmt>();
-    assert(doStmt && "missing DO statement");
-    const auto &loopControl =
-        std::get<std::optional<Fortran::parser::LoopControl>>(doStmt->t);
-    llvm::SmallVector<IncrementLoopInfo, 1> incrementLoopInfo;
-    const Fortran::parser::ScalarLogicalExpr *whileCondition = nullptr;
+    const auto &loopControl = std::get<1>(doStmt->t);
+    auto *preheaderBlock = doStmtEval.block;
+    auto *headerBlock =
+        unstructuredContext ? doStmtEval.localBlocks[0] : nullptr;
+    auto *bodyBlock = doStmtEval.lexicalSuccessor->block;
+    auto *exitBlock = doStmtEval.parentConstruct->constructExit->block;
+    llvm::SmallVector<IncrementLoopInfo, 4> incrementLoopInfo;
+    const Fortran::parser::ScalarLogicalExpr *whileCondition{};
     bool infiniteLoop = !loopControl.has_value();
     if (infiniteLoop) {
       assert(unstructuredContext && "infinite loop must be unstructured");
-      startBlock(doStmtEval.localBlocks[0]); // header block
-    } else if ((whileCondition =
-                    std::get_if<Fortran::parser::ScalarLogicalExpr>(
-                        &loopControl->u))) {
+      startBlock(headerBlock);
+    } else if ((whileCondition = std::get_if<1>(&loopControl->u))) {
```

Comment:
I assume there is more than 1 `Pa::ScalarLogicalExpr` in this variant...

Reply:
No, there is only one `ScalarLogicalExpr` in `LoopControl`.

If you are wondering why I switched the `std::get_if` argument from a type to an index here, it's part of a tradeoff.  There are 15 std::get_if calls in this function.  On one side, when they are an option (not always the case), symbolic types provide good information that indexes don't provide.  However, thanks to the long namespace prefixes (including `Fortran::`, which I think is unnecessary overkill), the symbolic types bloat the code with extra lines, as here, making the code harder to read.  The clincher for me here is that I often have a hard time reading even the symbolic types.  If I need to understand details, it is often necessary to reference parse-tree.h at the same time anyway, which reduces the benefit of the symbolic types.  `ScalarLogicalExpr` is one instance that is clear enough, but the context of looking for a `whileCondition` compensates.  So in this stretch of code, I thought the advantage of less physical code favored the index form, and I decided to use that uniformly everywhere.

If you prefer, I could change at least some of these (back) to symbolic types.

Reply:
The namespace issue is water under the bridge. Short names lost.

Please use the fully elaborated type names throughout rather than the integer constants. I believe that may be spelled out in the front-end coding covenants, rules, and restrictions; it's a rule that is followed everywhere possible.

Reply:
done